### ISSUE TITLE
Differences between the result of Triton Inference Server and mere GPU

### ISSUE THREAD

**Issue by nguyenhoa93** [NVIDIA/triton-inference-server#1578]
Hello,

I am quite new to Triton so please excuse me for naive questions.
I am using Triton Inference Server to inference for a 3-class classification and generate activation maps.

When I compare the results from Triton Inference Server and mere GPU, I noticed some different results:
* There is a slight difference between the probabilities. Not much, but still have some differences. (The speed of triton is excellent in comparison to normal GPU).
![inference-result](https://i.imgur.com/F74cuiA.png)
* Activation maps have a quite big difference (I have 3 models):
![grad1](https://i.imgur.com/fp33Ac0.png)
![grad2](https://i.imgur.com/5yryoS0.png)

My models were trained on Tensorflow 1.15. What I did:
1. I define the `gradient tensors` and `feature_map` tensor exactly the same way I did on GPU then saved model with `tf.saved_model.simple_save`
```
export_dir = "<export_dir>"
x = graph.get_tensor_by_name(<name_of_input_tensor>)
predict = graph.get_tensor_by_name("<name_of_output_tensor>")
final_conv = graph.get_tensor_by_name("<name_of_final_convolution_tensor>")

classes = [0,1,2]
def grads(classes,predict, final_conv):
    res = {}
    for idx in classes:
        one_hot = tf.sparse_to_dense(idx, [len(classes)],1.0)
        signal = tf.multiply(predict,one_hot)
        loss = tf.reduce_mean(signal)
        
        grads = tf.gradients(loss, final_conv)[0]
        
        norm_grads = tf.div(grads, tf.sqrt(tf.reduce_mean(tf.square(grads)))+tf.constant(1e-5))
        
        res[idx] = norm_grads
    
    return res
grads = grads(classes, predict, final_conv)

# Save model
tf.saved_model.simple_save(sess, export_dir=export_dir,
                          inputs={"input": x},
                          outputs={"output": predict,
                                   "final_conv": final_conv,
                                  "grad_2": grads[2],
                                  "grad_1": grads[1],
                                   "grad_0": grads[0]})
```

2. I preprocessed the images according to the design of the model before running 
3. My post-process to generate gradcam from `final_conv` values and gradient values from Triton server is also exactly the same way as I did on GPU.

My questions:
1. What are the possible causes of these differences?
2. How can I do to overcome this?

Thank you for your time.

**Comment by nguyenhoa93** [comment]
Hi @ethem-kinginthenorth 
You can see those values here: https://docs.google.com/spreadsheets/d/1On66xE-eqAzwXJSRhHoL9EUjfDHBsUp-yIDxrFJNIXc/edit?usp=sharing

**Comment by deadeyegoodwin** [comment]
Triton itself does not do anything with your input tensors, it simply feeds them to the model. Same for outputs. Usually a difference like this is because the preprocessing and/or post processing logic is not exactly identical between the 2 cases. Can  you compare the input tensors you are sending to Triton with what you are using in the "GPU" case (for the "GPU" case I assume you mean you are just calling the TF apis directly from python).

**Comment by nguyenhoa93** [comment]
> Triton itself does not do anything with your input tensors, it simply feeds them to the model. Same for outputs. Usually a difference like this is because the preprocessing and/or post processing logic is not exactly identical between the 2 cases. Can you compare the input tensors you are sending to Triton with what you are using in the "GPU" case (for the "GPU" case I assume you mean you are just calling the TF apis directly from python).

Thank you @deadeyegoodwin.

Indeed, when I printed input for Triton and input for TF APIs directly from python, I found them slightly different. And also post-process for GradCAM, I use `cv2.imwrite` to save the overlayed heatmap without changing the input image color from `RGB` to `BGR`. That's is the reason for the differences in the background.

I close the issue and try to handle the preprocess and post-process properly. Thank you for your time.

**Comment by sanjukta2706** [comment]
How can we access activations maps of intermediate layers in Triton?

**Comment by luvwinnie** [comment]
@nguyenhoa93 May I ask how did you process the GradCAM? Do you able to use triton inference server to generate the GradCAM heatmap?

**Comment by nguyenhoa93** [comment]
Hi @sanjukta2706 and @luvwinnie yes, it's possible.
You just need to create another model called gradcam model. The gradcam model inherits your trained model's weights and its outputs are the last convolutional layer and gradients. You can look at the example codes I posted above with tensorflow 1.15. After that you need a post-process step to generate heatmap. Besides, don't forget to make a config file for this model as you do for the normal model if necessary.

I think you can also make the model another way by directly output `last convolutional layer * class gradient`, so in the post-process, you don't need to do that. It depends on your needs.

**Comment by luvwinnie** [comment]
@nguyenhoa93 Thank you so much for replying! However it seems like it shows some errors when I use the tensorflow version 2's tf.compat.v1 API so save the savedModel. This is the [Tensorflow issue](https://github.com/tensorflow/tensorflow/issues/47292) I have opened and a [triton inference server issue](https://github.com/triton-inference-server/server/issues/2570)
